### PR TITLE
draft: Observer

### PR DESCRIPTION
**Pull request by Pnom**

### CODE REVIEW

File: src/obis-observer/commands/downlink/GetObisIdList.ts [hunk 1]
```diff
@@ -60,7 +68,9 @@ class GetObisIdList extends Command {
     constructor ( public parameters: IGetObisIdListParameters ) {
         super();
 
-        this.size = REQUEST_ID_SIZE + 1 + CommandBinaryBuffer.getObisSize(parameters.obis);
+        this.size = REQUEST_ID_SIZE
```
Comment (DarkPark) [inline src/obis-observer/commands/downlink/GetObisIdList.ts:71]: maybe in one line?

File: src/obis-observer/commands/downlink/GetObisIdList.ts [hunk 2]
```diff
@@ -91,7 +101,9 @@ class GetObisIdList extends Command {
 
         buffer.setUint8(requestId);
         buffer.setUint8(meterProfileId);
-        buffer.setObis(obis);
+        if ( obis ) {
```
Comment (DarkPark) [inline src/obis-observer/commands/downlink/GetObisIdList.ts:104]:
```suggestion

        if ( obis ) {
```
?

File: src/obis-observer/commands/uplink/GetMeterInfoResponse.ts [hunk 3]
```diff
@@ -63,8 +71,16 @@ class GetMeterInfoResponse extends Command {
     constructor ( public parameters: IGetMeterInfoResponseParameters ) {
         super();
 
-        // real size - request id byte + software version 2 bytes + hardware version 2 bytes + device name string size byte + string bytes
-        this.size = REQUEST_ID_SIZE + 1 + 1 + parameters.address.length;
+        let size = REQUEST_ID_SIZE;
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/GetMeterInfoResponse.ts:74]:
```suggestion
        let size = REQUEST_ID_SIZE;

```
?

File: src/obis-observer/commands/uplink/GetMeterInfoResponse.ts [hunk 4]
```diff
@@ -82,8 +98,12 @@ class GetMeterInfoResponse extends Command {
         const buffer = new CommandBinaryBuffer(data);
 
         const requestId = buffer.getUint8();
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/GetMeterInfoResponse.ts:100]: remove one empty line before and add one after

File: src/obis-observer/commands/downlink/GetObisIdList.ts (hunk 2 above)
Comment (Pnom) [reply src/obis-observer/commands/downlink/GetObisIdList.ts:109]: fixed

File: src/obis-observer/commands/uplink/GetMeterInfoResponse.ts (hunk 3 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/GetMeterInfoResponse.ts]: fixed

File: src/obis-observer/commands/uplink/GetMeterInfoResponse.ts (hunk 4 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/GetMeterInfoResponse.ts:113]: fixed

File: src/obis-observer/commands/uplink/GetObisInfoListResponse.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
+import Command, {TCommandExampleList} from '../../Command.js';
+import CommandBinaryBuffer, {REQUEST_ID_SIZE, ICommandParameters} from '../../CommandBinaryBuffer.js';
+import {UPLINK} from '../../constants/directions.js';
+
+
+type TObisInfo = {
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/GetObisInfoListResponse.ts:6]: it's better to make it an interface

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            isCompleted: true,
+            content: []
+        },
+        hex: {header: '12 02', body: '22 01'}
+    }
+];
+
+const commandSize = ( parameters: IReadMeterArchiveResponseParameters ): number => {
+    let size = COMMAND_HEADER_SIZE;
+
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:58]:
```suggestion
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+    // data - only body (without header)
+    static fromBytes ( data: Uint8Array ) {
+        const buffer = new CommandBinaryBuffer(data);
+        const requestId = buffer.getUint8();
+        const isCompleted = buffer.isEmpty ? true : buffer.getUint8() !== 0;
+        const content:Array<TArchiveContentPerDate> = [];
+
+        if ( buffer.isEmpty) {
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:139]:
```suggestion
        if ( buffer.isEmpty ) {
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            return new ReadMeterArchiveResponse({requestId, isCompleted, content});
+        }
+
+        while ( !buffer.isEmpty ) {
+            const dateContent:TArchiveContentPerDate = {
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:144]:
```suggestion
            const dateContent: TArchiveContentPerDate = {
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+
+    // data - only body (without header)
+    static fromBytes ( data: Uint8Array ) {
+        const buffer = new CommandBinaryBuffer(data);
+        const requestId = buffer.getUint8();
+        const isCompleted = buffer.isEmpty ? true : buffer.getUint8() !== 0;
+        const obisValueList:Array<IObisValueFloat> = [];
+
+        if ( buffer.isEmpty) {
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts:102]:
```suggestion
        if ( buffer.isEmpty ) {
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+ * ]);
+ * const command = ReadMeterArchiveWithDateResponse.fromBytes(commandBody);
+ *
+ * console.log(command.parameters);
+ * // output:
+ * {
+ *     requestId: 12,
+ *     isCompleted: true,
+ *     obisValueList: [{code: 8, content: 3.60}]
+ * },
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts:73]:
```suggestion
 * }
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 11]
```diff
@@ -0,0 +1,188 @@
+import Command, {TCommandExampleList} from '../../Command.js';
+import CommandBinaryBuffer, {REQUEST_ID_SIZE, ICommandParameters, IObisValueFloat, DATE_TIME_SIZE} from '../../CommandBinaryBuffer.js';
+import {UPLINK} from '../../constants/directions.js';
+import {TTime2000} from '../../../utils/time.js';
+import roundNumber from '../../../utils/roundNumber.js';
+
+
+type TArchiveContentPerDate = {
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:8]: interface if preferable

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        if ( obisValues.length !== 0 ) {
+            size += DATE_TIME_SIZE;
+            // 1 byte obis id + 4 byte float value for each obis value
+            size += (1 + 4) * obisValues.length;
+
+            datesInContent++;
+        }
+    }
+
+    if ( datesInContent !== 0) {
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:73]:
```suggestion
    if ( datesInContent !== 0 ) {
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+/**
+ * Uplink command.
+ *
+ * @example create command instance from command body hex dump
+ * ```js
+ * import ReadMeterArchiveResponse from 'jooby-codec/obis-observer/commands/uplink/ReadMeterArchiveResponse.js';
+ *
+ * const commandBody = new Uint8Array([
+ *     0c, 0x01, 0x1b, 0xb4, 0x0c, 0x60, 0x08, 0x3e, 0xcc, 0xcc, 0xcd, 0x00, 0x1b, 0xb4, 0x0c, 0x20, 0x08, 0x3e, 0x4c, 0xcc, 0xcd
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:90]:
```suggestion
 *     0x0c, 0x01, 0x1b, 0xb4, 0x0c, 0x60, 0x08, 0x3e, 0xcc, 0xcc, 0xcd, 0x00, 0x1b, 0xb4, 0x0c, 0x20, 0x08, 0x3e, 0x4c, 0xcc, 0xcd
```

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+ *         {
+ *             time2000: 464784480,
+ *             obisValueList: [{code: 8, content: 0.40}]
+ *         },
+ *         {
+ *             time2000: 464784416,
+ *             obisValueList: [{code: 8, content: 0.20}]
+ *         }
+ *     ]
+ * },
```
Comment (DarkPark) [inline src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts:110]:
```suggestion
 * }
```

File: src/obis-observer/commands/downlink/ReadMeterArchive.ts [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * @example
+ * ```js
+ * import ReadMeterArchive from 'jooby-codec/obis-observer/commands/downlink/ReadMeterArchive.js';
+ *
+ * const parameters = {
+ *     requestId: 3,
+ *     archiveType: 1,
+ *     startIndex: 4,
+ *     meterId: 2,
```
Comment (DarkPark) [inline src/obis-observer/commands/downlink/ReadMeterArchive.ts:45]:
```suggestion
 *     meterId: 2
```

**Review by DarkPark — CHANGES_REQUESTED** [review]
some changes are required

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 6 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 7 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 8 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts (hunk 9 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts (hunk 10 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveWithDateResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 12 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 13 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 14 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/downlink/ReadMeterArchive.ts (hunk 15 above)
Comment (Pnom) [reply src/obis-observer/commands/downlink/ReadMeterArchive.ts]: fixed

File: src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts (hunk 11 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/ReadMeterArchiveResponse.ts]: fixed

File: src/obis-observer/commands/uplink/GetObisInfoListResponse.ts (hunk 5 above)
Comment (Pnom) [reply src/obis-observer/commands/uplink/GetObisInfoListResponse.ts]: fixed

File: src/obis-observer/commands/downlink/UpdateImageWrite.ts [hunk 16]
```diff
@@ -51,7 +51,7 @@ const examples: TCommandExampleList = [
  *
  * // output command binary in hex representation
  * console.log(command.toHex());
- * // 2a 15 21 00 00 08 40 00 01 02 03 04 05 06 07 08 09 00 00 00 00 00 00
+ * // 32 15 21 00 00 08 40 00 01 02 03 04 05 06 07 08 09 00 00 00 00 00 00
```
Comment (DarkPark) [inline src/obis-observer/commands/downlink/UpdateImageWrite.ts:54]:
```suggestion
 * // 30 15 21 00 00 08 40 00 01 02 03 04 05 06 07 08 09 00 00 00 00 00 00
```

Comment (Pnom) [reply src/obis-observer/commands/downlink/UpdateImageWrite.ts]: fixed

**Review by DarkPark — APPROVED** [review]

**Pull request merged**